### PR TITLE
Deriving psf parameters via gradient descent

### PR DESCRIPTION
**Pull request by EshitaJoshi**
A few things changed this PR:

- Plotting moved to ```src/simtools/visualization/plot_psf.py```
- Random search algorithm for best parameters replaced by gradient descent minimisation of RMSD
- All elements of the parameter array are being used in the minimisation
-  Implemented two sample KS statistic as a sanity check for how close the data and simulation cumulative psf is
- Implemented a monte carlo analysis of errors with the ```--monte_carlo_analysis``` flag that runs the simulations 500 times with the same parameters and outputs the variation on rmsd, d80, ks statistic, and p value. Doesn't give an error on the parameters, but the psf.
- Added ```PSFcurve_data_v2.ecsv``` and changed to using astropy tables

Note that since the RMSD value is calculated on cumulative data which is normalised from 0-1, it can be thought as a percentage difference between data and simulation. So an RMSD of 0.01 means that on average, the simulated and measured data differ by 1 percentage across all radii.

An example of the cumulative pdf plots is here: [psf_gradient_descent_plots_LSTN-01.pdf](https://github.com/user-attachments/files/22274845/psf_gradient_descent_plots_LSTN-01.pdf)

How the gradient descent is implemented:

- Initial RMSD is calculated between data and simulation
- Parameters are perturbed in one direction by adding an epsilon to each element in the list
- New RMSD calculated, this is used to find the gradient
- Parameters are moved in the direction of negative gradient as ```new_value = old_value - learning_rate*gradient```
- If there is a rejected step reduce the learning rate and try again
- If there are more than three consecutive rejected steps maybe we are in a local minimum so we increase the learning rate and try again
- Max iterations capped at 100 to avoid infinite tries, otherwise iterate until threshold RMSD is reached
- Threshold RMSD and learning rate can be set as CLI arguments

To run gradient descent:

`python src/simtools/applications/derive_psf_parameters.py --site North --telescope LSTN-01 --model_version 6.0.0 --plot_all --test --rmsd_threshold 0.008 --learning_rate 0.005 --data tests/resources/PSFcurve_data_v2.ecsv --write_psf_parameters 2>&1 | grep -E "(Initial RMSD|Accepted step|Step rejected|iteration)"`

gradient_descent_convergence_LSTN-01

To get monte carlo errors for rmsd run with the ```--monte_carlo_analysis``` flag:

monte_carlo_uncertainty_LSTN-01_rmsd

To get the monte carlo errors for ks statistic just run with the ```--monte_carlo_analysis --ks_statistic``` flags:

monte_carlo_uncertainty_LSTN-01_ks

### CODE REVIEW

File: src/simtools/ray_tracing/psf_parameter_optimisation.py [hunk 1]
```diff
@@ -651,21 +192,8 @@ def write_tested_parameters_to_file(results, best_pars, best_d80, output_dir, te
...
-    ----------
-    best_pars : dict
-        Dictionary with PSF parameter names as keys and values as lists
-
-    Returns
-    -------
-    dict
-        Dictionary with same keys but values converted to astropy quantities with units
-    """
+    """Add proper astropy units to PSF parameters based on their schemas."""
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/psf_parameter_optimisation.py:195]: Removed

File: src/simtools/ray_tracing/psf_parameter_optimisation.py [hunk 2]
```diff
@@ -311,338 +103,87 @@ def run_psf_simulation(
...
-        plt.grid(True, alpha=0.3)
+    d_type = {"names": (RADIUS_CM, CUMULATIVE_PSF), "formats": ("f8", "f8")}
+    data = np.zeros(len(table), dtype=d_type)
 
-        plot_file_name = f"tune_psf_{tel_model.name}_best_params_{key}.pdf"
-        plot_file = output_dir.joinpath(plot_file_name)
-        visualize.save_figure(plt, plot_file, log_title=f"D80 vs off-axis ({key})")
+    data[RADIUS_CM] = table["radius_mm"].quantity.to(u.cm).value
+    data[CUMULATIVE_PSF] = table["cumulative_psf"]
+    data[CUMULATIVE_PSF] /= np.max(np.abs(data[CUMULATIVE_PSF]))  # Normalize to max = 1.0
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply src/simtools/ray_tracing/psf_parameter_optimisation.py:158]: Yes!

File: src/simtools/applications/derive_psf_parameters.py [hunk 3]
```diff
@@ -63,47 +62,45 @@
...
+            --model_version 6.0.0 --data tests/resources/PSFcurve_data_v2.ecsv --plot_all --test
 
     Run with parameter export:
 
     .. code-block:: console
 
-        simtools-derive-psf-parameters --site North --telescope LSTN-01 \\
-            --model_version 6.0.0 \\
-            --data tests/resources/PSFcurve_data_v2.txt --write_psf_parameters
+        simtools-derive-psf-parameters --site North --telescope LSTN-01 --model_version 6.0.0 \\
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply src/simtools/applications/derive_psf_parameters.py:89]: Yes, this is a good place.

File: src/simtools/ray_tracing/psf_parameter_optimisation.py [hunk 4]
```diff
@@ -682,111 +210,712 @@ def _add_units_to_psf_parameters(best_pars):
...
+                args_dict,
+                perturbed_params,
+                data_to_plot,
+                radius,
+                pdf_pages=None,
+                is_best=False,
+                use_ks_statistic=use_ks_statistic,
+            )
+            param_gradients.append((perturbed_rmsd - current_rmsd) / epsilon)
+        except (ValueError, RuntimeError):
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/psf_parameter_optimisation.py:277]: Hm, no not really. Gradient 0 implies the simulation completed without errors and that we're at a turning point. Maybe I can use NaN instead. If it's ok I will add this to the todo list in the next PR and look into it in more detail there?

File: src/simtools/ray_tracing/psf_parameter_optimisation.py [hunk 5]
```diff
@@ -311,338 +103,87 @@ def run_psf_simulation(
...
-            f"align_horizontal={best_pars['mirror_align_random_horizontal'][0]:.4f}\n"
-            f"align_vertical={best_pars['mirror_align_random_vertical'][0]:.4f}\n"
-        )
-        plt.xlabel("Off-axis Angle (degrees)")
-        plt.ylabel("D80 (cm)" if key == "d80_cm" else "D80 (degrees)")
-        plt.ylim(bottom=0)
-        plt.xticks(rotation=45)
-        plt.xlim(0, max_offset)
-        plt.grid(True, alpha=0.3)
+    d_type = {"names": (RADIUS_CM, CUMULATIVE_PSF), "formats": ("f8", "f8")}
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/psf_parameter_optimisation.py:153]: Done

File: src/simtools/ray_tracing/psf_parameter_optimisation.py [hunk 6]
```diff
@@ -682,111 +210,712 @@ def _add_units_to_psf_parameters(best_pars):
...
-    """
-    # Generate parameter sets
-    all_parameters = []
-    mrra_0, mfr_0, mrra2_0, mar_0 = get_previous_values(tel_model)
-
-    n_runs = args_dict.get("n_runs")
-    generate_random_parameters(
-        all_parameters, n_runs, args_dict, mrra_0, mfr_0, mrra2_0, mar_0, tel_model
-    )
+def _calculate_param_gradient(
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/psf_parameter_optimisation.py:237]: Done

File: src/simtools/ray_tracing/psf_parameter_optimisation.py [hunk 7]
```diff
@@ -682,111 +210,712 @@ def _add_units_to_psf_parameters(best_pars):
...
+    site_model,
+    args_dict,
+    current_params,
+    data_to_plot,
+    radius,
+    current_rmsd,
+    epsilon=0.0005,
+    use_ks_statistic=False,
+):
+    """Calculate numerical gradient of RMSD with respect to parameters."""
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/psf_parameter_optimisation.py:294]: Done

File: src/simtools/ray_tracing/psf_parameter_optimisation.py [hunk 8]
```diff
@@ -682,111 +210,712 @@ def _add_units_to_psf_parameters(best_pars):
...
+            param_values,
+            epsilon,
+            use_ks_statistic,
+        )
+
+    return gradients
+
+
+def apply_gradient_step(current_params, gradients, learning_rate):
+    """Move parameters in the direction of negative gradient."""
```
[the remark this answers is not in the document]
Comment (EshitaJoshi) [reply src/simtools/ray_tracing/psf_parameter_optimisation.py:316]: Done

**Comment by EshitaJoshi** [comment]
@GernotMaier thank you for the thorough review! Please let me know if the changes are alright.

File: tests/resources/PSFcurve_data_v2.ecsv [hunk 9]
```diff
@@ -0,0 +1,51 @@
+# %ECSV 1.0
+# ---
+# datatype:
+# - {name: radius_mm, unit: 'mm', datatype: float64, description: radial distance}
+# - {name: differential_value, unit: '', datatype: float64, description: differential PSF value}
+# - {name: cumulative_psf, unit: '', datatype: float64, description: cumulative PSF value}
+# meta: !!omap
+# - comments: ['PSF taken at Zd = ~20[deg], with Vega']
+# schema: astropy-2.0
+radius_mm differential_value cumulative_psf
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply tests/resources/PSFcurve_data_v2.ecsv]: Important was to get the unit out of the column name.

For the second point, what is used in simtools in the analysis? This particular file is not so important (it is an example for software development), but we need to be clear on this in future and either enforce a certain definition or be able to digest different definitions (but it should be indicated in the metadata of the ecsv file). 

I would say that this is beyond this PR, so leave it as it is.
